src/state_machine/coordinator/fire.rs

```diff
@@ -0,0 +1,1296 @@
+        // XXX this needs to be key_ids for v1 and signer_ids for v2
+        let party_ids = self
+            .public_nonces
+            .values()
+            .flat_map(|pn| pn.key_ids.clone())
+            .collect::<Vec<u32>>();
+        let nonces = self
+            .public_nonces
+            .values()
+            .flat_map(|pn| pn.nonces.clone())
+            .collect::<Vec<PublicNonce>>();
+        let (_, R) = compute::intermediate(&self.message, &party_ids, &nonces);
+
+        R
+    }
+}
+
+impl<Aggregator: AggregatorTrait> StateMachine<State, Error> for Coordinator<Aggregator> {
+    fn move_to(&mut self, state: State) -> Result<(), Error> {
+        self.can_move_to(&state)?;
+        self.state = state;
+        Ok(())
+    }
+
+    fn can_move_to(&self, state: &State) -> Result<(), Error> {
+        let prev_state = &self.state;
+        let accepted = match state {
+            State::Idle => true,
+            State::DkgPublicDistribute => {
+                prev_state == &State::Idle
+                    || prev_state == &State::DkgPublicGather
+                    || prev_state == &State::DkgEndGather
+            }
+            State::DkgPublicGather => {
+                prev_state == &State::DkgPublicDistribute || prev_state == &State::DkgPublicGather
+            }
+            State::DkgPrivateDistribute => prev_state == &State::DkgPublicGather,
+            State::DkgEndGather => prev_state == &State::DkgPrivateDistribute,
+            State::NonceRequest(is_taproot, merkle_root) => {
+                prev_state == &State::Idle
+                    || prev_state == &State::DkgEndGather
+                    || prev_state == &State::SigShareGather(*is_taproot, *merkle_root)
+            }
+            State::NonceGather(is_taproot, merkle_root) => {
+                prev_state == &State::NonceRequest(*is_taproot, *merkle_root)
+                    || prev_state == &State::NonceGather(*is_taproot, *merkle_root)
+            }
+            State::SigShareRequest(is_taproot, merkle_root) => {
+                prev_state == &State::NonceGather(*is_taproot, *merkle_root)
+            }
+            State::SigShareGather(is_taproot, merkle_root) => {
+                prev_state == &State::SigShareRequest(*is_taproot, *merkle_root)
+                    || prev_state == &State::SigShareGather(*is_taproot, *merkle_root)
+            }
+        };
+        if accepted {
+            debug!("state change from {:?} to {:?}", prev_state, state);
+            Ok(())
+        } else {
+            Err(Error::BadStateChange(format!(
+                "{:?} to {:?}",
+                prev_state, state
+            )))
+        }
+    }
+}
+
+impl<Aggregator: AggregatorTrait> CoordinatorTrait for Coordinator<Aggregator> {
+    /// Create a new coordinator
+    fn new(config: Config) -> Self {
+        Self {
+            aggregator: Aggregator::new(config.num_keys, config.threshold),
+            config,
+            current_dkg_id: 0,
```

Comment:
This isn't really that important. But wonder if we could clean this up to have a struct called "RoundInfo" or something since a lot of these are common between the different types of coordinators. Is a nit though and may actually make things uglier so feel free to ignore.

Reply:
Yeah there needs to be a way of encapsulating common code.  Too bad Rust is so terrible for OO semantics.